api/v1alpha3/cluster_types.go

```diff
@@ -178,9 +178,9 @@ type APIEndpoint struct {
 	Port int32 `json:"port"`
 }
 
-// IsZero returns true if host and the port are zero values.
+// IsZero returns true if either the host or the port is zero value.
 func (v APIEndpoint) IsZero() bool {
```

Comment:
hello, i don't think this change is appropriate, since it does a change in behavior.

side note, IsZero is a misleading name for a function that works on non-numerical types.
after this change the function name becomes more misleading.

> Current Endpoint.IsZero() check will allow empty Host with non-empty Port to pass the check (vice versa). This will cause issue if Infra provider have partial endpoint information with either of them is empty. The specific case we saw is with CAPV when user need to manually configure endpoint for kube-vip based HA.

this seems like the responsibility of a different validation function that ensures that both fields are set.

Reply:
Agree that the name might be misleading. All the controllers (bootstrap/kcp) relies on this check to decide whether it should proceed or keep waiting. We could add a new function and change those controllers, then this one will be only used in conversion/validation.  @ncdc

Reply:
Should we hold this until v0.4? Just to point out here as well (after our discussion on slack) the Host and Port field are required in open api validation, this is really an extra check.

Reply:
I don't 100% agree that `IsZero` is misleading, as the zero value for a struct means all its fields have their default initialized (zero) values. This function is currently doing what I'd expect based on what a Go struct's zero value means.

However, I do agree this function may be misleading to the end user, who is probably calling it to check if the APIEndpoint has been set, and having it return false (i.e. it has been set) when the host is empty but the port is filled in seems wrong.

Let's maybe add a new function, something like `IsComplete`, that returns true iff both host and port are nonzero?

Reply:
+1, I'd say we're going into semantic validation

Reply:
+1 for IsComplete

The open api validation doesn't seem to prevent this case, since both default zero values and a patch with only port is accepted by the apiserver.

Reply:
Let's use `IsValid` which is consistent with other parts of the codebase. And yeah, upon further validation the fields are required, but nothing stops an empty string (we should fix this in v1alpha4)

Reply:
With the current changes are there any remaining uses of IsZero? If not, we should likely mark the method as deprecated.

Reply:
it's used in conversion.go, to decide whether apiendpoint is set.